File: sim/golden_models/models/ClkRstModel.cpp
#include "golden_models/models/ClkRstModel.h"

#include <filesystem>

#include "golden_models/core/TestVectorIO.h"

namespace fpd::sim {

ClkRstModel::ClkRstModel(uint64_t afe_clk_hz) : afe_clk_hz_(afe_clk_hz) {
    reset();
}

void ClkRstModel::reset() {
    phase_acc_aclk_ = 0;
    phase_acc_mclk_ = 0;
    rst_ext_n_ = false;
    afe_type_sel_ = 0;
    clk_afe_ = false;
    clk_aclk_ = false;
    clk_mclk_ = false;
    pll_locked_ = false;
    rst_ff1_ = false;
    rst_ff2_ = false;
    lock_counter_ = 0;
    cycle_count_ = 0;
}

void ClkRstModel::step() {
    if (!rst_ext_n_) {
        phase_acc_aclk_ = 0;
        phase_acc_mclk_ = 0;
        clk_afe_ = false;
        clk_aclk_ = false;
        clk_mclk_ = false;
        pll_locked_ = false;
        rst_ff1_ = false;
        rst_ff2_ = false;
        lock_counter_ = 0;
    } else {
        const uint64_t phase_step_aclk = (afe_clk_hz_ << 32U) / kSysClkHz;
        const uint64_t phase_step_mclk = (kMclkHz << 32U) / kSysClkHz;
        phase_acc_aclk_ = (phase_acc_aclk_ + phase_step_aclk) & 0xFFFFFFFFULL;
        phase_acc_mclk_ = (phase_acc_mclk_ + phase_step_mclk) & 0xFFFFFFFFULL;
        clk_aclk_ = ((phase_acc_aclk_ >> 31U) & 0x1ULL) != 0ULL;
        clk_mclk_ = ((phase_acc_mclk_ >> 31U) & 0x1ULL) != 0ULL;
        clk_afe_ = (afe_type_sel_ == 2U) ? clk_mclk_ : clk_aclk_;

        if (!pll_locked_) {
            ++lock_counter_;
            if (lock_counter_ >= 16U) {
                pll_locked_ = true;
            }
        }

        rst_ff1_ = pll_locked_;
        rst_ff2_ = rst_ff1_;
    }

    ++cycle_count_;
}

void ClkRstModel::set_inputs(const SignalMap& inputs) {
    rst_ext_n_ = (GetScalar(inputs, "rst_ext_n", rst_ext_n_ ? 1U : 0U) & 0x1U) != 0U;
    afe_type_sel_ = static_cast<uint8_t>(GetScalar(inputs, "afe_type_sel", afe_type_sel_) & 0x3U);
}

SignalMap ClkRstModel::get_outputs() const {
    return {
        {"clk_afe", clk_afe_ ? 1U : 0U},
        {"clk_aclk", clk_aclk_ ? 1U : 0U},
        {"clk_mclk", clk_mclk_ ? 1U : 0U},
        {"clk_sys_out", 1U},
        {"rst_sync_n", rst_ff2_ ? 1U : 0U},
        {"pll_locked", pll_locked_ ? 1U : 0U},
    };
}

std::vector<Mismatch> ClkRstModel::compare(const SignalMap& rtl_outputs) const {
    return CompareSignalMaps(cycle(), get_outputs(), rtl_outputs);
}

void ClkRstModel::generate_vectors(const std::string& output_dir) {
    std::filesystem::create_directories(output_dir);

    TestVectorFile vector_file;
    vector_file.module_name = "clk_rst_mgr";
    vector_file.spec_name = "SPEC-FPD-001";
    vector_file.signal_inputs = {"rst_ext_n", "afe_type_sel"};
    vector_file.signal_outputs = {"pll_locked", "rst_sync_n", "clk_afe"};

    reset();

    for (int i = 0; i < 20; ++i) {
        set_inputs({{"rst_ext_n", i < 2 ? 0U : 1U}, {"afe_type_sel", 0U}});
        step();
        vector_file.vectors.push_back(TestVector{
            cycle(),
            {{"rst_ext_n", i < 2 ? 0U : 1U}, {"afe_type_sel", 0U}},
            get_outputs(),
        });
    }

    WriteHexVectors(vector_file, output_dir + "/clk_rst_mgr.hex");
}

}  // namespace fpd::sim

File: sim/viewer/src/FpdSimViewer/Models/ClkRstModel.cs
using FpdSimViewer.Models.Core;

namespace FpdSimViewer.Models;

public sealed class ClkRstModel : GoldenModelBase
{
    private const ulong kSysClkHz = 100000000UL;
    private const ulong kMclkHz = 32000000UL;

    private readonly ulong _afeClkHz;

    private ulong _phaseAccAclk;
    private ulong _phaseAccMclk;
    private bool _rstExtN;
    private byte _afeTypeSel;
    private bool _clkAfe;
    private bool _clkAclk;
    private bool _clkMclk;
    private bool _pllLocked;
    private bool _rstFf1;
    private bool _rstFf2;
    private uint _lockCounter;

    public ClkRstModel(ulong afeClkHz = 10000000UL)
    {
        _afeClkHz = afeClkHz;
        Reset();
    }

    public override void Reset()
    {
        _phaseAccAclk = 0UL;
        _phaseAccMclk = 0UL;
        _rstExtN = false;
        _afeTypeSel = 0;
        _clkAfe = false;
        _clkAclk = false;
        _clkMclk = false;
        _pllLocked = false;
        _rstFf1 = false;
        _rstFf2 = false;
        _lockCounter = 0U;
        CycleCount = 0;
    }

    public override void Step()
    {
        if (!_rstExtN)
        {
            _phaseAccAclk = 0UL;
            _phaseAccMclk = 0UL;
            _clkAfe = false;
            _clkAclk = false;
            _clkMclk = false;
            _pllLocked = false;
            _rstFf1 = false;
            _rstFf2 = false;
            _lockCounter = 0U;
        }
        else
        {
            var phaseStepAclk = (_afeClkHz << 32) / kSysClkHz;
            var phaseStepMclk = (kMclkHz << 32) / kSysClkHz;
            _phaseAccAclk = (_phaseAccAclk + phaseStepAclk) & 0xFFFFFFFFUL;
            _phaseAccMclk = (_phaseAccMclk + phaseStepMclk) & 0xFFFFFFFFUL;
            _clkAclk = ((_phaseAccAclk >> 31) & 0x1UL) != 0UL;
            _clkMclk = ((_phaseAccMclk >> 31) & 0x1UL) != 0UL;
            _clkAfe = _afeTypeSel == 2U ? _clkMclk : _clkAclk;

            if (!_pllLocked)
            {
                _lockCounter++;
                if (_lockCounter >= 16U)
                {
                    _pllLocked = true;
                }
            }

            // Intentionally matches the C++ golden model's same-cycle propagation.
            _rstFf1 = _pllLocked;
            _rstFf2 = _rstFf1;
        }

        CycleCount++;
    }

    public override void SetInputs(SignalMap inputs)
    {
        _rstExtN = (SignalHelpers.GetScalar(inputs, "rst_ext_n", _rstExtN ? 1U : 0U) & 0x1U) != 0U;
        _afeTypeSel = (byte)(SignalHelpers.GetScalar(inputs, "afe_type_sel", _afeTypeSel) & 0x3U);
    }

    public override SignalMap GetOutputs()
    {
        var rstSync = _rstFf2 ? 1U : 0U;
        return new SignalMap
        {
            ["clk_afe"] = _clkAfe ? 1U : 0U,
            ["clk_aclk"] = _clkAclk ? 1U : 0U,
            ["clk_mclk"] = _clkMclk ? 1U : 0U,
            ["clk_sys_out"] = 1U,
            ["pll_locked"] = _pllLocked ? 1U : 0U,
            ["rst_sync"] = rstSync,
            ["rst_sync_n"] = rstSync,
        };
    }
}
